src/hotspot/os/linux/os_linux.cpp

```diff
@@ -3696,12 +3702,26 @@ static void set_coredump_filter(CoredumpFilterBit bit) {
 
 static size_t _large_page_size = 0;
 
-void warn_no_large_pages_configured() {
+static void warn_no_large_pages_configured() {
   if (!FLAG_IS_DEFAULT(UseLargePages)) {
     log_warning(pagesize)("UseLargePages disabled, no large pages configured and available on the system.");
   }
 }
 
+static void warn_thp_not_configured() {
+  if (FLAG_IS_DEFAULT(UseTransparentHugePages)) {
+    // Don't warn if user didn't set the flag
+    return;
+  }
+
+  if (UseZGC && HugePages::thp_requested() && HugePages::supports_shmem_thp()) {
+    // ZGC uses shared memory THPs; don't warn if those are configured
+    return;
+  }
+
+  log_warning(pagesize)("UseTransparentHugePages disabled, transparent huge pages are not supported by the operating system.");
```

Comment:
Would it be not clearer to define when to warn, as we do in warn_no_large_pages?

Related to that, should we not warn if ZGC and +shmemthp configured but -anonymous thp? I am not sure the heap is the only part of the JVM that uses THP, and other parts would still use anon THP, or? E.g. Code heap.

Also, maybe a better message for the poor admin that tries to setup. E.g.:

```
bool requires_shmem_thp = UseTHP + UseZGC
bool requires_anon_thp = UseTHP
bool off = false;

if (requires_shmem && !shmem configured) 
  (log_warning "Shmem thp are not supported. Set /sys/kernel/mm/transparent_hugepage/shmem_enabled  to advise to support shmem thp")
  off = true;

if (requires_anonthp && !anon_thp configured) 
  (log_warning "anonymous Thp are not supported. Set /sys/kernel/mm/transparent_hugepage/enabled  to madvise")
  off = true;

if (off) 
  UseTHP = 0
  log_warning(UseTHP disabled (see previous messages)
```

 if ZGC and !supports shmemthp or

Reply:
I think this is okay, what do you think? Too many messages?